pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java

```diff
@@ -155,8 +155,14 @@ public static MessageId fromByteArrayWithTopic(byte[] data, TopicName topicName)
 
         MessageId messageId;
         if (idData.hasBatchIndex()) {
-            messageId = new BatchMessageIdImpl(idData.getLedgerId(), idData.getEntryId(), idData.getPartition(),
-                idData.getBatchIndex(), idData.getBatchSize(), BatchMessageAcker.newAcker(idData.getBatchSize()));
+            if (idData.hasBatchSize()) {
```

Comment:
Is there any case in that we will have a batch index but no batch size?
Maybe we should throw an exception due to it should be the case that corrupted data is provided.

Reply:
The `batch_size` field was introduced from https://github.com/apache/pulsar/pull/8659, before that, the `batch_index` field already existed. For example, the 2.7.1 client could send such a message id, see https://github.com/apache/pulsar/issues/18395#issuecomment-1309864662.

Other languages clients might also sent a message id that has the `batch_index` field but no `batch_size` field.

Reply:
And when I revisited the existing code, there are many places that set the `batch_index` field but doesn't set the `batch_size` field. The `batch_size` field is easily to be ignored because it's not considered in `equals` and `hashCode` methods. BTW, I'm not sure whether the default value of the `batch_index` field might increase the cases that `batch_index` is set while `batch_size` is not set. IMO, adding default values to the optional fields increases the chance of such an error. We should be careful for new fields if it's optional.

Reply:
Got it.

Reply:
@codelipenghui FWIW we even have already done it for `fromByteArray` (as this patch modifies).

Reply:
@BewareMyPower So you may simplify the code as `fromByteArray` did:

```java
            if (idData.hasBatchSize()) {
                messageId = new BatchMessageIdImpl(idData.getLedgerId(), idData.getEntryId(), idData.getPartition(),
                    idData.getBatchIndex(), idData.getBatchSize(), BatchMessageAcker.newAcker(idData.getBatchSize()));
            } else {
                messageId = new BatchMessageIdImpl(idData.getLedgerId(), idData.getEntryId(), idData.getPartition(),
                    idData.getBatchIndex());
            }
```

That is, a simpler constructor. Or factor it out to a helper method.

Reply:
I would like to refactor these two classes in another PR. IMO, BatchMessageIdImpl should be deprecated and I'm not willing to add new public APIs to it. The design of BatchMessageIdImpl is bad. We should add the optional `batch_size` and `batch_index` fields to MessageIdImpl instead. If we followed the design of BatchMessageIdImpl, when `batch_size` was introduced, we should add another class like BatchMessageIdImplWithSize that inherits the BatchMessageIdImpl and add the `batch_size` field as well as the acker.